src/rmc-compiler/src/main.rs

```diff
@@ -0,0 +1,227 @@
+                .short("V")
+                .long("--version")
+                .help("Gets underlying rustc version."),
+        )
+        .arg(
+            Arg::with_name("rustc-options")
+                .help("Arguments to be passed down to rustc.")
+                .multiple(true)
+                .takes_value(true),
+        )
+}
+
+/// Main function. Configure arguments and run the compiler.
+fn main() -> Result<(), &'static str> {
+    let matches = parser().get_matches();
+
+    // Initialize the logger.
+    init_env_logger("RMC_LOG");
+
+    // Generate rustc args.
+    let rustc_args = generate_rustc_args(&matches);
+
+    // Configure queries.
+    let mut queries = QueryDb::default();
+    queries
+        .set_symbol_table_passes(matches.values_of_lossy("symbol-table-passes").unwrap_or(vec![]));
+    queries.set_emit_vtable_restrictions(matches.is_present("restrict-vtable-fn-ptrs"));
+
+    // Configure and run compiler.
+    let mut callbacks = RmcCallbacks {};
+    install_ice_hook();
+    let mut compiler = RunCompiler::new(&rustc_args, &mut callbacks);
+    if matches.is_present("goto-c") {
+        compiler.set_make_codegen_backend(Some(Box::new(move |_cfg| {
+            rustc_codegen_rmc::GotocCodegenBackend::new(&Rc::new(queries))
+        })));
+    }
+    compiler.run().or(Err("Failed to compile crate."))
+}
+
+/// Empty struct since we don't support any callbacks yet.
+struct RmcCallbacks {}
+
+/// Use default function implementations.
+impl Callbacks for RmcCallbacks {}
+
+/// Generate the arguments to pass to rustc_driver.
+fn generate_rustc_args(args: &ArgMatches) -> Vec<String> {
+    let mut gotoc_args = rustc_gotoc_flags(&args.value_of("rmc-lib").unwrap());
+    let mut rustc_args = vec![String::from("rustc")];
+    if args.is_present("goto-c") {
+        rustc_args.append(&mut gotoc_args);
+    }
+
+    if args.is_present("rustc-version") {
+        rustc_args.push(String::from("--version"))
+    }
+
+    rustc_args.append(
+        &mut args
+            .values_of("rustc-options")
+            .unwrap_or(clap::Values::default())
+            .map(|s| s.to_string())
+            .collect::<Vec<_>>(),
+    );
+    let sysroot = sysroot_path(args.value_of("sysroot")).unwrap();
+    rustc_args.push(String::from("--sysroot"));
+    rustc_args.push(sysroot.to_string_lossy().to_string());
+    tracing::info!(?rustc_args, "Compile");
+    rustc_args
+}
+
+/// Try to generate the rustup toolchain path.
+fn toolchain_path(home: Option<String>, toolchain: Option<String>) -> Option<PathBuf> {
+    match (home, toolchain) {
+        (Some(home), Some(toolchain)) => {
+            Some([home, String::from("toolchains"), toolchain].iter().collect::<PathBuf>())
+        }
+        _ => None,
+    }
+}
+
+/// Get the sysroot, following the order bellow:
+/// - "--sysroot" command line argument
+/// - compile-time environment
+///    - $SYSROOT
+///    - $RUSTUP_HOME/toolchains/$RUSTUP_TOOLCHAIN
```

Comment:
are these environment variables typically provided by cargo? Or is this just another option we're creating to supply them?

Reply:
Yes, they are usually provided by cargo. This is similar to code from clippy and Miri.